octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,343 @@
+
+    def _apply(self, plan):
+        """Required function of manager.py to actually apply a record change.
+
+            :param plan: Contains the zones and changes to be made
+            :type  plan: octodns.provider.base.Plan
+
+            :type return: void
+        """
+        desired = plan.desired
+        changes = plan.changes
+
+        self.log.debug('_apply: zone=%s, len(changes)=%d', desired.name,
+                       len(changes))
+
+        # Get gcloud zone, or create one if none existed before.
+        gcloud_zone = self._get_gcloud_zone(desired.name, create=True)
+
+        gcloud_changes = gcloud_zone.changes()
+
+        for change in changes:
+            class_name = change.__class__.__name__
+            _rrset_func = getattr(
+                self, '_rrset_for_{}'.format(change.record._type))
+
+            if class_name in 'Create':
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Delete':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.record))
+            elif class_name == 'Update':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.existing))
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.new))
+            else:
+                raise RuntimeError('Change type "{}" for change "{!s}" '
+                                   'is none of "Create", "Delete" or "Update'
+                                   .format(class_name, change))
+
+        gcloud_changes.create()
+        i = 1
+        while gcloud_changes.status != 'done':
+            self.log.debug("Waiting for changes to complete")
+            time.sleep(i)
+            gcloud_changes.reload()
+            if i < 30:
+                i += 2
+
+    def _create_gcloud_zone(self, dns_name):
+        """Creates a google cloud ManagedZone with dns_name, and zone named
+            derived from it. calls .create() method and returns it.
+
+            :param dns_name: fqdn of zone to create
+            :type  dns_name: str
+
+            :type return: new google.cloud.dns.ManagedZone
+        """
+        # Zone name must begin with a letter, end with a letter or digit,
+        # and only contain lowercase letters, digits or dashes
+        zone_name = re.sub("[^a-z0-9-]", "",
+                           dns_name[:-1].replace('.', "-"))
+        # make sure that the end result did not end up wo leading letter
+        if re.match('[^a-z]', zone_name[0]):
+            # I cannot think of a situation where a zone name derived from
+            # a domain name would'nt start with leading letter and thereby
+            # violate the constraint, however if such a situation is
+            # encountered, add a leading "a" here.
+            zone_name = "a%s" % zone_name
```

Comment:
Hrm. Not 100% sure either as it's isn't something I've dug into. A quick search seems turned up https://stackoverflow.com/questions/7111881/what-are-the-allowed-characters-in-a-subdomain and https://en.wikipedia.org/wiki/Domain_name.

> Domain names may be formed from the set of alphanumeric ASCII characters (a-z, A-Z, 0-9), but characters are case-insensitive. In addition the hyphen is permitted if it is surrounded by characters, digits or hyphens, although it is not to start or end a label. Labels are always separated by the full stop (period) character in the textual name representation.

So I think all of this can be omitted/ignored and assumed to be well behaved. If we do want to enforce this it'd probably make sense to add as rules to the Zone class (in a separate future PR)

Reply:
OK makes sense.